gnovm/pkg/gnolang/op_expressions.go

```diff
@@ -204,16 +204,41 @@ func (m *Machine) doOpRef() {
 func (m *Machine) doOpTypeAssert1() {
 	m.PopExpr()
 	// pop type
-	t := m.PopValue().GetType()
+	t := m.PopValue().GetType() // type being asserted
+
 	// peek x for re-use
-	xv := m.PeekValue(1)
-	xt := xv.T
+	xv := m.PeekValue(1) // value result / value to assert
+	xt := xv.T           // underlying value's type
+
+	// xt may be nil, but we need to wait to return because the value of xt that is set
+	// will depend on whether we are trying to assert to an interface or concrete type.
+	// xt can be nil in the case where recover can't find a panic to recover from and
+	// returns a bare TypedValue{}.
 
 	if t.Kind() == InterfaceKind { // is interface assert
+		if xt == nil {
+			// TODO: default panic type?
+			ex := fmt.Sprintf("nil doesn't implement %s", t.String())
+			m.Panic(typedString(ex))
+			return
+		}
+
 		if it, ok := baseOf(t).(*InterfaceType); ok {
+			// An interface type assertion on a value that doesn't have a concrete base
+			// type should always fail.
+			if _, ok := baseOf(xt).(*InterfaceType); ok {
+				// TODO: default panic type?
+				ex := fmt.Sprintf(
+					"non-concrete %s doesn't implement %s",
+					xt.String(),
+					it.String())
+				m.Panic(typedString(ex))
+				return
+			}
+
```

Comment:
Yes there are. Like in go, there are untyped nil values and typed nil values. `xt == nil` evaluates to true when `xv` is an uptyped nil value; `xv.T` is nil. As far as I can tell, this can only happen two ways:
1) the value that is returned by `recover` if there is no panic to recover
2) below in the `doOpTypeAssert2` method when a type assertion fails -- `*xv = TypedValue{}`

So an example of a case where `t` is `*InterfaceType` and the underlying value (`xv`) is not nil is in the `type35a.gno` file test: https://github.com/gnolang/gno/blob/347db5de435cdf836e043e550c880d6d5b907858/gnovm/tests/files/type35a.gno